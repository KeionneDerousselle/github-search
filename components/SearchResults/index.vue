<template>
  <div
    ref="results"
    class="results"
    v-bind="$attrs">
    <ul
      v-for="result in results"
      :id="`page-${result.page}`"
      :key="result.id">
      <user-list-item
        v-for="user in result.users"
        :id="`search-result-${user.id}`"
        :key="user.id"
        :user="user"
        :selected="selectedUser && selectedUser.id === user.id"
        class="search__result"
        @click="handleUserSelected(user)" />
    </ul>
  </div>
</template>
<script>
import UserListItem from '@/components/UserListItem'
import { mapActions, mapGetters } from 'vuex'

export default {
  components: {
    UserListItem
  },

  data: () => ({
    maxScrollPosition: 0,
    fetchingTheNextPage: false,
    currentScrollCancellation: null,
    selectedUser: null
  }),

  computed: {
    ...mapGetters('users', [
      'results',
      'currentSearchTerm',
      'resultsPerPage',
      'currentPage'
    ])
  },

  mounted() {
    this.addScrollListener(this.$refs.results)
  },

  beforeDestroy() {
    this.removeScrollListener(this.$refs.results)
  },

  methods: {
    ...mapActions('users', [
      'search',
      'setPage'
    ]),

    addScrollListener(el) {
      el.addEventListener('scroll', this.handleResultsScrolled)
    },

    removeScrollListener(el) {
      el.removeEventListener('scroll', this.handleResultsScrolled)
    },

    handleResultsScrolled() {
      this.handleScrollHelper(this.$refs.results)
    },

    async handleScrollHelper(el) {
      // when height of content under the view window (the bottom of the scroll: (this.$refs.results.scrollHeight - this.$refs.results.scrollTop))
      // is less than 1.5 pages: (this.$refs.results.offsetHeight * 1.5)
      if ((el.scrollHeight - el.scrollTop) < (el.offsetHeight * 1.5)) {
        // if the bottom of the content is not the same as the max position we've scrolled to
        if ((el.scrollHeight >= this.maxScrollPosition) && !this.fetchingTheNextPage) {
          this.fetchingTheNextPage = true
          this.setMaxScollHeight(el)

          await this.search({
            searchTerm: this.currentSearchTerm,
            page: this.currentPage + 1,
            resultsPerPage: this.resultsPerPage
          }).finally(() => {
            this.fetchingTheNextPage = false
          })
        }
      }

      this.updateScrolledPage(el)
    },

    setMaxScollHeight(el) {
      this.maxScrollPosition = el.scrollHeight
    },

    updateScrolledPage(el) {
      const calculatedPage = this.calculateScrolledToPage(el)

      if (calculatedPage !== this.currentPage) {
        return this.setPage(calculatedPage)
      }
    },

    calculateScrolledToPage(el) {
      const children = el.children
      let previousMax = 0
      let currentMax = 0
      const scrollTop = el.scrollTop

      // for every ul.page el in #results
      for (let i = 0; i < children.length; i++) {
        // console.log('hi')
        // get the height of that page
        currentMax += children[i].clientHeight
        // if the scroll position is between the start of the page and the end of the page, return that we are on that page
        if (scrollTop >= previousMax && scrollTop <= currentMax) return i + 1
        // else, add the height of this page to the height of all the pages we've already checked, (reset the start looking position)
        previousMax = currentMax
      }
    },

    handleUserSelected(user) {

    }
  }
}
</script>
