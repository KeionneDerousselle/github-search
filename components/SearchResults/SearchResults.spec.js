import flushPromises from 'flush-promises'
import SearchResults from '.'

describe('Search Results', () => {
  let wrapper

  const createWrapper = ({ store = {} } = {}) => {
    const { users, ...storeModules } = store

    return mountPreMocked(SearchResults, {
      store: {
        users: {
          getters: {
            results: jest.fn().mockReturnValue([]),
            currentSearchTerm: jest.fn().mockReturnValue(''),
            resultsPerPage: jest.fn().mockReturnValue(25),
            currentPage: jest.fn().mockReturnValue(1),
            userDetailsByUsername: jest.fn().mockReturnValue(jest.fn().mockReturnValue()),
            ...(users?.getters || {})
          },

          actions: {
            search: jest.fn().mockResolvedValue(),
            setPage: jest.fn().mockResolvedValue(),
            get: jest.fn().mockResolvedValue(),
            ...(users?.actions || {})
          }
        },
        ...storeModules
      }
    })
  }

  describe('when there are search results to be displayed', () => {
    let results

    beforeAll(() => {
      results = [
        {
          id: 1,
          page: 1,
          users: [
            { id: 1, login: 'user1' },
            { id: 2, login: 'user2' },
            { id: 3, login: 'user3' }
          ]
        },
        {
          id: 2,
          page: 2,
          users: [
            { id: 4, login: 'user4' },
            { id: 5, login: 'user5' },
            { id: 6, login: 'user6' }
          ]
        }
      ]

      wrapper = createWrapper({
        store: {
          users: {
            getters: {
              results: jest.fn().mockReturnValue(results),
              resultsPerPage: jest.fn().mockReturnValue(1),
              currentPage: jest.fn().mockReturnValue(1)
            }
          }
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should display a search result for each user', () => {
      results.forEach(result => {
        const resultList = wrapper.get(`#page-${result.page}`)

        expect(resultList.element).toBeVisible()

        result.users.forEach(user => {
          const userSearchResult = wrapper.get(`#search-result-${user.id}`)

          expect(userSearchResult.element).toBeVisible()
        })
      })
    })
  })

  describe('when there are no search results to be displayed', () => {
    beforeAll(() => {
      wrapper = createWrapper()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should display the search results container', () => {
      expect(wrapper.get('.results').element).toBeVisible()
    })

    it('should not display any search results ', () => {
      expect(() => wrapper.get('.search__result')).toThrow()
    })
  })

  describe('when the adding the scroll listener', () => {
    let mockedResultsScrollBox

    beforeAll(() => {
      wrapper = createWrapper()

      mockedResultsScrollBox = {
        addEventListener: jest.fn()
      }

      wrapper.vm.addScrollListener(mockedResultsScrollBox)
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should add an event listener to the element that is passed in', () => {
      expect(mockedResultsScrollBox.addEventListener).toHaveBeenCalledWith('scroll', wrapper.vm.handleResultsScrolled)
    })
  })

  describe('when removing the scroll listener', () => {
    let mockedResultsScrollBox

    beforeAll(() => {
      wrapper = createWrapper()

      mockedResultsScrollBox = {
        removeEventListener: jest.fn()
      }

      wrapper.vm.removeScrollListener(mockedResultsScrollBox)
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should remove an event listener to the element that is passed in', () => {
      expect(mockedResultsScrollBox.removeEventListener).toHaveBeenCalledWith('scroll', wrapper.vm.handleResultsScrolled)
    })
  })

  describe('handleResultsScrolled', () => {
    let handleScrollHelperSpy

    beforeAll(() => {
      wrapper = createWrapper()
      handleScrollHelperSpy = jest.spyOn(wrapper.vm, 'handleScrollHelper')

      wrapper.vm.handleResultsScrolled()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should remove an event listener to the element that is passed in', () => {
      expect(handleScrollHelperSpy).toHaveBeenCalledWith(wrapper.vm.$refs.results)
    })
  })

  describe('when scrolling and the scrollable content under the results box is greater than 1.5 pages', () => {
    let mockedResultsScrollBox
    let mockedSearch
    let updateScrolledPageSpy

    beforeAll(async () => {
      mockedSearch = jest.fn().mockResolvedValue()
      wrapper = createWrapper({
        store: {
          users: {
            actions: {
              search: mockedSearch
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollHeight: 770,
        scrollTop: 0,
        offsetHeight: 500
      }

      updateScrolledPageSpy = jest.spyOn(wrapper.vm, 'updateScrolledPage')
      updateScrolledPageSpy.mockImplementation(() => {})

      await wrapper.vm.handleScrollHelper(mockedResultsScrollBox)
    })

    afterAll(() => {
      updateScrolledPageSpy.mockRestore()
      wrapper.destroy()
    })

    it('should not fetch the next page of results', () => {
      expect(mockedSearch).not.toHaveBeenCalled()
    })

    it('should not set the max scroll position to the height of the scrollable content', () => {
      expect(wrapper.vm.maxScrollPosition).not.toBe(mockedResultsScrollBox.scrollHeight)
    })

    it('should update the current page to the scrolled page', () => {
      expect(updateScrolledPageSpy).toBeCalled()
    })
  })

  describe('when scrolling and the scrollable content under the results box is less than 1.5 pages', () => {
    let mockedResultsScrollBox
    let mockedSearch
    let updateScrolledPageSpy
    let currentSearchTerm

    beforeAll(async () => {
      currentSearchTerm = 'test'
      mockedSearch = jest.fn().mockResolvedValue()

      wrapper = createWrapper({
        store: {
          users: {
            getters: {
              currentSearchTerm: jest.fn().mockReturnValue(currentSearchTerm)
            },
            actions: {
              search: mockedSearch
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollHeight: 500,
        scrollTop: 300,
        offsetHeight: 500
      }

      updateScrolledPageSpy = jest.spyOn(wrapper.vm, 'updateScrolledPage')
      updateScrolledPageSpy.mockImplementation(() => {})

      await wrapper.vm.handleScrollHelper(mockedResultsScrollBox)
      await flushPromises()
    })

    afterAll(() => {
      updateScrolledPageSpy.mockRestore()
      wrapper.destroy()
    })

    it('should fetch the next page of results', () => {
      expect(mockedSearch).toHaveBeenCalledWith(expect.any(Object), {
        searchTerm: currentSearchTerm,
        page: wrapper.vm.currentPage + 1,
        resultsPerPage: wrapper.vm.resultsPerPage
      })
    })

    it('should set the max scroll position to the height of the scrollable content', () => {
      expect(wrapper.vm.maxScrollPosition).toBe(mockedResultsScrollBox.scrollHeight)
    })

    it('should update the current page to the scrolled page', () => {
      expect(updateScrolledPageSpy).toHaveBeenCalled()
    })
  })

  describe('when scrolling and the scrollable content under the results box is less than 1.5 pages, but we are already fetching the next page', () => {
    let mockedResultsScrollBox
    let mockedSearch
    let updateScrolledPageSpy
    let currentSearchTerm

    beforeAll(async () => {
      currentSearchTerm = 'test'
      mockedSearch = jest.fn().mockResolvedValue()

      wrapper = createWrapper({
        store: {
          users: {
            getters: {
              currentSearchTerm: jest.fn().mockReturnValue(currentSearchTerm)
            },
            actions: {
              search: mockedSearch
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollHeight: 500,
        scrollTop: 300,
        offsetHeight: 500
      }

      wrapper.setData({
        fetchingTheNextPage: true
      })

      updateScrolledPageSpy = jest.spyOn(wrapper.vm, 'updateScrolledPage')
      updateScrolledPageSpy.mockImplementation(() => { })

      await wrapper.vm.handleScrollHelper(mockedResultsScrollBox)
      await flushPromises()
    })

    afterAll(() => {
      updateScrolledPageSpy.mockRestore()
      wrapper.destroy()
    })

    it('should not fetch the next page of results', () => {
      expect(mockedSearch).not.toHaveBeenCalled()
    })

    it('should not set the max scroll position to the height of the scrollable content', () => {
      expect(wrapper.vm.maxScrollPosition).not.toBe(mockedResultsScrollBox.scrollHeight)
    })

    it('should update the current page to the scrolled page', () => {
      expect(updateScrolledPageSpy).toHaveBeenCalled()
    })
  })

  describe('when updating the scrolled to page, and the scrolled page is greater than the current page', () => {
    let mockedSetPage
    let mockedResultsScrollBox

    beforeAll(async () => {
      mockedSetPage = jest.fn().mockResolvedValue()
      wrapper = createWrapper({
        store: {
          users: {
            actions: {
              setPage: mockedSetPage
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollTop: 300,
        children: [ { clientHeight: 150 }, { clientHeight: 300 } ]
      }

      await wrapper.vm.updateScrolledPage(mockedResultsScrollBox)
    })

    afterAll(() => {
      mockedSetPage.mockReset()
      wrapper.destroy()
    })

    it('should set the current page to the scrolled to page ', () => {
      expect(mockedSetPage).toHaveBeenCalledWith(expect.any(Object), 2)
    })
  })

  describe('when updating the scrolled to page, and the scrolled page is less than the current page', () => {
    let mockedSetPage
    let mockedResultsScrollBox

    beforeAll(async () => {
      mockedSetPage = jest.fn().mockResolvedValue()
      wrapper = createWrapper({
        store: {
          users: {
            getters: {
              currentPage: jest.fn().mockReturnValue(2)
            },

            actions: {
              setPage: mockedSetPage
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollTop: 100,
        children: [ { clientHeight: 150 }, { clientHeight: 300 } ]
      }
      await wrapper.vm.updateScrolledPage(mockedResultsScrollBox)
    })

    afterAll(() => {
      mockedSetPage.mockReset()
      wrapper.destroy()
    })

    it('should set the current page to the scrolled to page ', () => {
      expect(mockedSetPage).toHaveBeenCalledWith(expect.any(Object), 1)
    })
  })

  describe('when updating the scrolled to page, and the scrolled page is the same as the current page', () => {
    let mockedSetPage
    let mockedResultsScrollBox

    beforeAll(async () => {
      mockedSetPage = jest.fn().mockResolvedValue()
      wrapper = createWrapper({
        store: {
          users: {
            actions: {
              setPage: mockedSetPage
            }
          }
        }
      })

      mockedResultsScrollBox = {
        scrollTop: 150,
        children: [ { clientHeight: 150 }, { clientHeight: 300 } ]
      }

      await wrapper.vm.updateScrolledPage(mockedResultsScrollBox)
    })

    afterAll(() => {
      mockedSetPage.mockReset()
      wrapper.destroy()
    })

    it('should not set the current page to the scrolled to page ', () => {
      expect(mockedSetPage).not.toBeCalled()
    })
  })
})
