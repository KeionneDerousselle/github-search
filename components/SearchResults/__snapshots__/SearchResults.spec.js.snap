// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Results when there are no search results to be displayed should render as expected 1`] = `<div class="results"></div>`;

exports[`Search Results when there are search results to be displayed should render as expected 1`] = `
<div class="results">
  <ul id="page-1">
    <li id="search-result-1" class="search__result user-list-item"><img alt="Profile picture for user1" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user1</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user1's GitHub Account in a new window.</span></a>
    </li>
    <li id="search-result-2" class="search__result user-list-item"><img alt="Profile picture for user2" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user2</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user2's GitHub Account in a new window.</span></a>
    </li>
    <li id="search-result-3" class="search__result user-list-item"><img alt="Profile picture for user3" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user3</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user3's GitHub Account in a new window.</span></a>
    </li>
  </ul>
  <ul id="page-2">
    <li id="search-result-4" class="search__result user-list-item"><img alt="Profile picture for user4" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user4</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user4's GitHub Account in a new window.</span></a>
    </li>
    <li id="search-result-5" class="search__result user-list-item"><img alt="Profile picture for user5" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user5</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user5's GitHub Account in a new window.</span></a>
    </li>
    <li id="search-result-6" class="search__result user-list-item"><img alt="Profile picture for user6" class="user-list-item__avatar">
      <div class="user-list-item__details">
        <!----> <span class="user-list-item__details__login">user6</span>
      </div>
      <!----> <a target="_blank" rel="noopener" class="user-list-item__link"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true" class="user-list-item__link__icon">
          <path d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"></path>
          <path d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"></path>
        </svg> <span class="sr-only"> Opens link to user6's GitHub Account in a new window.</span></a>
    </li>
  </ul>
</div>
`;
